optuna/storages/_journal/redis.py

```diff
@@ -0,0 +1,66 @@
+import json
+import time
+from typing import Any
+from typing import Dict
+from typing import List
+
+from optuna._experimental import experimental_class
+from optuna._imports import try_import
+from optuna.storages._journal.base import BaseJournalLogStorage
+
+
+with try_import() as _imports:
+    import redis
+
+
+@experimental_class("3.1.0")
+class JournalRedisStorage(BaseJournalLogStorage):
+    """Redis storage class for Journal log backend.
+
+    Args:
+        url:
+            URL of the redis storage, password and db are optional. (ie: redis://localhost:6379)
+
+    """
+
+    def __init__(self, url: str) -> None:
+
+        _imports.check()
+
+        self._url = url
+        self._redis = redis.Redis.from_url(url)
+
+    def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
+
+        max_log_number_bytes = self._redis.get("log_number")
+        if max_log_number_bytes is None:
+            return []
+        max_log_number = int(max_log_number_bytes)
+
+        logs = []
+        for log_number in range(log_number_from, max_log_number + 1):
+            sleep_secs = 0.1
+            while True:
+                log_bytes = self._redis.get(self._key_log_id(log_number))
+                if log_bytes is not None:
+                    break
+                time.sleep(sleep_secs)
+                sleep_secs = min(sleep_secs * 2, 10)
+            log = log_bytes.decode("utf-8")
+            try:
+                logs.append(json.loads(log))
+            except json.JSONDecodeError as err:
+                if log_number != max_log_number:
+                    raise err
+        return logs
+
+    def append_logs(self, logs: List[Dict[str, Any]]) -> None:
+
+        self._redis.setnx("log_number", -1)
+        for log in logs:
+            log_number = self._redis.incr("log_number", 1)
+            self._redis.set(self._key_log_id(log_number), json.dumps(log))
```

Comment:
If `incr log_number` succeeds and `set log:{log_number} ...` fails,  the following `read_logs` calls will never be returned. How about making this operation more atomic by using Redis scripting feature?
(Please note that we need to install `fakeredis[lua]` to run unittests. See https://github.com/cunla/fakeredis-py for details.)

```suggestion
            self._redis.eval(
                "local i = redis.call('incr', 'log_number') "
                "redis.call('set', string.format('log:%d', i), KEYS[1])",
                1,
                json.dumps(log),
            )
```

Furthermore,  as written in the documentation, `register_script` makes this logic more efficient.

> In practice, use the object returned by register_script. This function exists purely for Redis API completion.
> https://redis-py.readthedocs.io/en/stable/commands.html?highlight=register_script#redis.commands.core.CoreCommands.eval

Reply:
Sorry, I've just found the following notation.

> Important: to ensure the correct execution of scripts, both in standalone and clustered deployments, all names of keys that a script accesses must be explicitly provided as input key arguments. The script should only access keys whose names are given as input arguments. Scripts should never access keys with programmatically-generated names or based on the contents of data structures stored in the database.
> https://redis.io/docs/manual/programmability/eval-intro/

We cannot list all keys that the script accesses since "log:{log_number}" is dynamically generated.

Reply:
Not sure, but perhaps we should dynamically generate script source code per its needs. According to the documentation, it is an anti-pattern due to script cache considerations though.

```suggestion
            self._redis.eval(
                "local i = redis.call('incr', 'log_number') "
                f"redis.call('set', string.format('log:%d', i), {json.dumps(log)})",
                0
            )
```